### PR TITLE
Friends

### PR DESCRIPTION
**Pull request by adamerdemer**
- Adds friendships
- Added stub user profile to show friends
- Added find friends link
- Users are unique
- Cannot add yourself 
- Sign out button disappears after signing out (conditional)

### CODE REVIEW

File: app/controllers/friendships_controller.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
+class FriendshipsController < ApplicationController
+  def create
+    @friendship = current_user.friendships.build(friend_id: params[:friend_id])
```
Comment (enonnai) [inline app/controllers/friendships_controller.rb]: If signed out, trying to add a friend returns a NoMethodError on line 3. Would probably need to return a flash notice saying you need to sign up/ log in to add friends.

Comment (sophieklm) [reply app/controllers/friendships_controller.rb]: good spot - will update...

File: app/models/friendship.rb [hunk 2]
```diff
@@ -0,0 +1,11 @@
+class Friendship < ActiveRecord::Base
+  belongs_to :user
+  belongs_to :friend, :class_name => "User"
+  validates_uniqueness_of :friend, scope: :user
+  validate :check_user
+  def check_user
```
Comment (anaalta) [inline app/models/friendship.rb:6]: When you try to add yourself, you get the same Error adding friend error

File: app/views/users/index.html.erb [hunk 3]
```diff
@@ -0,0 +1,7 @@
+<ul>
+  <% @users.each do |user| %>
+    <p>
+      <%= user.email %><%= link_to "Add Friend", friendships_path(:friend_id => user), :method => :post %>
+    </p>
+  <% end %>
+</ul>
```
Comment (anaalta) [inline app/views/users/index.html.erb]: Maybe add a if current_user (linked to the fact that you can see a list of emails and add friend buttons after you signed out)

**Review by anaalta — CHANGES_REQUESTED** [review]
looks awesome! Great work! see comments for changes. also, instead of stubbing, maybe use Factory Girl gem?

File: app/controllers/friendships_controller.rb [hunk 4]
```diff
@@ -1,14 +1,17 @@
...
+      if @friendship.save
+        flash[:notice] = "Added friend"
+      else
+        friend = @friendship.friend
+        flash[:notice] = "Error Adding Friend"
+      end
     else
-      friend = @friendship.friend
-      flash[:notice] = "Error Adding Friend"
+      flash[:notice] = "You're not signed in!"
```
Comment (enonnai) [inline app/controllers/friendships_controller.rb]: Functionality works just fine! I would just change the wording into something like 'Sign up or Log in to start adding friends' :)

File: app/models/friendship.rb (hunk 2 above)
Comment (sophieklm) [reply app/models/friendship.rb:6]: I'm finding it difficult to implement flash errors to get the check user error to render, so I might have to hand over to someone else.

**Comment by JessicaBarclay** [comment]
Looks Awesome, well done guys getting this going 💯

**Comment by JessicaBarclay** [comment]
The only issue I could spot is that User email addresses appear on the homepage when signed out? But this may be resolved when merged with other features that the Newsfeed guys worked on.

:shipit:

**Comment by chrisjmit** [comment]
:shipit: possibly worth raising an issue for the add friend link displaying on home page when signed out, and a to do for friendship approvals so users can be added to both friend lists on approval. congrats!

**Review by chrisjmit — APPROVED** [review]

**Comment by anaalta** [comment]
:shipit: 
isn't Factory Girl magical? :)

**Review by anaalta — APPROVED** [review]

File: app/models/friendship.rb (hunk 2 above)
Comment (sophieklm) [reply app/models/friendship.rb:6]: Managed to update this

**Comment by enonnai** [comment]
Great job guys! Would be nice to change 'Users' with 'Friends' to keep it consistent with the rest of the instructions and to have an error saying 'This user is already a friend' when trying to add a friend who has been added already, but for me it's a :shipit: !!!

**Review by enonnai — APPROVED** [review]

**Pull request merged**